tokio/src/sync/rwlock/owned_read_guard.rs

```diff
@@ -0,0 +1,145 @@
+use crate::sync::rwlock::RwLock;
+use std::fmt;
+use std::mem;
+use std::mem::ManuallyDrop;
+use std::ops;
+use std::sync::Arc;
+
+/// Owned RAII structure used to release the shared read access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`read_owned`] method on
+/// [`RwLock`].
+///
+/// [`read_owned`]: method@crate::sync::RwLock::read_owned
+/// [`RwLock`]: struct@crate::sync::RwLock
+pub struct OwnedRwLockReadGuard<T: ?Sized, U: ?Sized = T> {
+    // ManuallyDrop allows us to destructure into this field without running the destructor.
+    pub(super) lock: ManuallyDrop<Arc<RwLock<T>>>,
+    pub(super) data: *const U,
+}
```

Comment:
Can you add a `PhantomData<T>` field to better instruct dropcheck? Same for the others.

Reply:
I don't actually know for sure whether it is necessary, but I know that it is definitely correct dropcheck-wise with the `PhantomData`.